src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs

```diff
@@ -44,30 +38,27 @@ public override async ValueTask<T> GetAssetAsync<T>(Checksum checksum, Cancellat
         {
             Contract.ThrowIfTrue(checksum == Checksum.Null);
             if (_assetCache.TryGetAsset<T>(checksum, out var asset))
-            {
                 return asset;
-            }
 
-            using (Logger.LogBlock(FunctionId.AssetService_GetAssetAsync, Checksum.GetChecksumLogInfo, checksum, cancellationToken))
-            {
-                var value = await RequestAssetAsync(checksum, cancellationToken).ConfigureAwait(false);
+            using var pooledObject = SharedPools.Default<HashSet<Checksum>>().GetPooledObject();
+            var checksums = pooledObject.Object;
+            checksums.Add(checksum);
 
-                // Attempt to add the value to the cache.  If someone else beat us, we'll return their value.
-                return (T)_assetCache.GetOrAdd(checksum, value);
-            }
+            await this.SynchronizeAssetsAsync(checksums, cancellationToken).ConfigureAwait(false);
+
+            return GetRequiredAsset<T>(checksum);
         }
 
         public async ValueTask<ImmutableArray<ValueTuple<Checksum, T>>> GetAssetsAsync<T>(HashSet<Checksum> checksums, CancellationToken cancellationToken)
         {
-            // this only works when caller wants to get same kind of assets at once
-
             // bulk synchronize checksums first
             var syncer = new ChecksumSynchronizer(this);
             await syncer.SynchronizeAssetsAsync(checksums, cancellationToken).ConfigureAwait(false);
 
+            // this will be fast since we actually synchronized the checksums above.
             using var _ = ArrayBuilder<ValueTuple<Checksum, T>>.GetInstance(checksums.Count, out var list);
             foreach (var checksum in checksums)
-                list.Add(ValueTuple.Create(checksum, await GetAssetAsync<T>(checksum, cancellationToken).ConfigureAwait(false)));
+                list.Add(ValueTuple.Create(checksum, GetRequiredAsset<T>(checksum)));
```

Comment:
a lot of these calls to "get the asset asycnhronously" became "get teh required asset" since they are called directly after the call to actually synchronize the asset.  This allows us to make stronger assertions that the data must be present, and that it's the prior syncronization call taht is fetching the items, and not this potential GetASsetAsync call that is doing it.